File: src/main/session.ts
/**
 * Session snapshot + crash recovery — MAIN only.
 *
 * 1:1 port of UWP `SessionManager` / `SessionUtility` (E:\Projects\Notepads):
 *   - A versioned `NotepadsSessionData.json` (Version==1) is written to the app's
 *     userData root, dirty-checked against the last serialized JSON so we only
 *     touch disk when the session actually changed (UWP `_lastSessionJsonStr`).
 *   - Per-tab content is mirrored into extension-less backup files under
 *     `{userData}/BackupFiles/`: `{editorId}-LastSaved` (the on-disk snapshot)
 *     and `{editorId}-Pending` (unsaved edits, only when the tab is dirty).
 *   - A corrupt session JSON is treated like UWP's `SessionDataCorruptedException`:
 *     extension-less backup files are renamed to `{name}-Corrupted.txt` and
 *     `loadLast` returns null (no crash, fresh start).
 *   - PA-4 FutureAccessList substitute: stored ABSOLUTE PATHS are re-validated on
 *     load via `revalidatePath` (fs.stat). A missing/renamed file marks the tab
 *     `unavailable:true` with its `filePath` PRESERVED (mirrors UWP's
 *     `GetItemAsync` try/catch silent-skip), distinct from an untitled buffer.
 *
 * The renderer NEVER touches fs/path — all of this lives behind IPC (PA-8). The
 * internal `_backups` sidecar (underscore-prefixed) is stripped before any
 * snapshot crosses IPC back to the renderer.
 */

import { mkdir, readFile, writeFile, readdir, rename, rm, stat } from 'node:fs/promises';
import { join } from 'node:path';
import { app } from 'electron';
import type { Result, SessionSnapshot, SessionTab } from '../shared/ipc-contract.js';
import { revalidatePath } from './file-io.js';

/** UWP `SessionUtility.SessionMetaDataFileDefaultName`. */
const SESSION_FILE_NAME = 'NotepadsSessionData.json';
/** UWP `SessionUtility.BackupFolderDefaultName`. */
const BACKUP_FOLDER_NAME = 'BackupFiles';

/** Per-tab backup payload kept ONLY in MAIN; never serialized into the JSON. */
interface TabBackup {
  /** The last on-disk content for the tab (UWP `{id}-LastSaved`). */
  lastSaved?: string;
  /** Unsaved edits, present only when the tab is dirty (UWP `{id}-Pending`). */
  pending?: string;
}

/**
 * Internal session shape: the contract snapshot plus an underscore sidecar that
 * carries per-editor backup text. The sidecar is the substitute for UWP's
 * separate backup-file references and is ALWAYS stripped before IPC.
 */
interface InternalSnapshot extends SessionSnapshot {
  /** editorId -> backup text. Underscore-prefixed: stripped before IPC. */
  _backups?: Record<string, TabBackup>;
}

function errMsg(e: unknown): string {
  return e instanceof Error ? e.message : String(e);
}

/**
 * Resolve the userData root. Honors the e2e override (`NOTEPADS_E2E_USERDATA`,
 * also applied via `app.setPath` in index.ts) so a scripted kill→restart hits
 * the SAME session JSON + backups. Falls back to Electron's userData path.
 */
function userDataRoot(): string {
  const override = process.env['NOTEPADS_E2E_USERDATA'];
  if (override && override.length > 0) return override;
  return app.getPath('userData');
}

function sessionFilePath(): string {
  return join(userDataRoot(), SESSION_FILE_NAME);
}

function backupFolderPath(): string {
  return join(userDataRoot(), BACKUP_FOLDER_NAME);
}

function lastSavedBackupName(editorId: string): string {
  return `${editorId}-LastSaved`;
}

function pendingBackupName(editorId: string): string {
  return `${editorId}-Pending`;
}

/**
 * Last serialized session JSON, mirroring UWP `_lastSessionJsonStr`. Used to
 * dirty-check snapshots so the 7s timer only writes when state actually changed.
 */
let lastSessionJsonStr: string | null = null;

/**
 * Serialize a snapshot to the on-disk JSON string. The `_backups` sidecar is
 * excluded — the JSON only ever holds the contract surface (Version==1, tabs,
 * activeEditorId), matching UWP's `NotepadsSessionDataV1`.
 */
function serializeForDisk(data: SessionSnapshot): string {
  const ordered = {
    version: 1 as const,
    tabs: data.tabs,
    activeEditorId: data.activeEditorId
  };
  return JSON.stringify(ordered, null, 2);
}

/**
 * Persist a session snapshot. Dirty-checked against the last written JSON
 * (case-insensitive, UWP `StringComparison.OrdinalIgnoreCase`); when unchanged
 * we return `{ written: false }` without touching disk. On a real write we (1)
 * write the versioned JSON to userData root, (2) write per-tab LastSaved/Pending
 * backups, (3) delete orphaned backups for editors no longer in the session.
 */
export async function snapshot(data: InternalSnapshot): Promise<Result<{ written: boolean }>> {
  try {
    const sessionJsonStr = serializeForDisk(data);
    if (
      lastSessionJsonStr !== null &&
      lastSessionJsonStr.toLowerCase() === sessionJsonStr.toLowerCase()
    ) {
      return { ok: true, data: { written: false } };
    }

    await mkdir(backupFolderPath(), { recursive: true });
    await writeFile(sessionFilePath(), sessionJsonStr, 'utf8');
    await writeBackups(data);
    await deleteOrphanedBackups(data.tabs);

    lastSessionJsonStr = sessionJsonStr;
    return { ok: true, data: { written: true } };
  } catch (e) {
    return { ok: false, error: errMsg(e) };
  }
}

/**
 * Write per-tab backup files from the `_backups` sidecar. LastSaved is written
 * whenever provided; Pending is written only when the tab is dirty and pending
 * text exists, mirroring UWP's `LastSavedSnapshot.Content != GetText()` guard.
 */
async function writeBackups(data: InternalSnapshot): Promise<void> {
  const backups = data._backups;
  if (!backups) return;
  const folder = backupFolderPath();
  const byId = new Map(data.tabs.map((t) => [t.editorId, t]));

  for (const [editorId, backup] of Object.entries(backups)) {
    if (backup.lastSaved != null) {
      await writeFile(join(folder, lastSavedBackupName(editorId)), backup.lastSaved, 'utf8');
    }
    const tab = byId.get(editorId);
    const pendingPath = join(folder, pendingBackupName(editorId));
    if (tab?.isModified && backup.pending != null) {
      await writeFile(pendingPath, backup.pending, 'utf8');
    } else {
      // Tab is clean (or has no pending edits): drop any stale pending backup so
      // recovery never re-surfaces resolved edits. rm(force) is a no-op if absent.
      await rm(pendingPath, { force: true });
    }
  }
}

/**
 * Delete extension-less backup files whose editorId is no longer present in the
 * live session (UWP `DeleteOrphanedBackupFilesAsync`). Files WITH an extension
 * (e.g. `*-Corrupted.txt`) are skipped — they are user-facing recovery dumps.
 */
async function deleteOrphanedBackups(tabs: SessionTab[]): Promise<void> {
  const folder = backupFolderPath();
  const liveNames = new Set<string>();
  for (const t of tabs) {
    liveNames.add(lastSavedBackupName(t.editorId));
    liveNames.add(pendingBackupName(t.editorId));
  }

  let entries: string[];
  try {
    entries = await readdir(folder);
  } catch {
    return; // folder absent -> nothing to prune
  }

  for (const name of entries) {
    if (name.includes('.')) continue; // skip files with extension
    if (liveNames.has(name)) continue; // skip known-live backups
    await rm(join(folder, name), { force: true });
  }
}

/**
 * Load the last persisted session.
 *
 * Three recovery cases (UWP `LoadLastSessionAsync` + `RecoverBackupFilesAsync`):
 *   (a) no session file               -> return null (fresh start).
 *   (b) session file, no pending text -> restore tabs from LastSaved content.
 *   (c) session file with pending     -> restore tabs from Pending content.
 *
 * Corrupt JSON is treated as UWP's `SessionDataCorruptedException`: rename the
 * extension-less backups to `{name}-Corrupted.txt` and return null.
 *
 * PA-4: every tab.filePath is re-validated via `revalidatePath` (fs.stat). A
 * missing/renamed file sets `unavailable:true` with `filePath` PRESERVED.
 */
export async function loadLast(): Promise<Result<SessionSnapshot | null>> {
  let raw: string;
  try {
    raw = await readFile(sessionFilePath(), 'utf8');
  } catch {
    return { ok: true, data: null }; // case (a): no file
  }

  let parsed: SessionSnapshot;
  try {
    parsed = JSON.parse(raw) as SessionSnapshot;
    if (parsed?.version !== 1 || !Array.isArray(parsed.tabs)) {
      throw new Error('Unexpected session shape');
    }
  } catch {
    // Corrupt JSON -> rename backups, drop the bad session file, fresh start.
    await renameCorruptedBackups();
    return { ok: true, data: null };
  }

  try {
    const tabs: SessionTab[] = [];
    for (const tab of parsed.tabs) {
      tabs.push(await revalidateTab(tab));
    }
    lastSessionJsonStr = serializeForDisk(parsed);
    return { ok: true, data: { version: 1, tabs, activeEditorId: parsed.activeEditorId } };
  } catch (e) {
    return { ok: false, error: errMsg(e) };
  }
}

/**
 * PA-4 re-validation for a single tab. Untitled buffers (filePath:null) pass
 * through untouched. A backed file that fails fs.stat is marked unavailable with
 * its path PRESERVED so the UI can offer relocate/save-as. The content itself is
 * recovered from the backup files by the renderer; the path stays authoritative.
 */
async function revalidateTab(tab: SessionTab): Promise<SessionTab> {
  if (!tab.filePath) {
    const { unavailable: _unavailable, ...rest } = tab;
    return rest;
  }
  const res = await revalidatePath(tab.filePath);
  const exists = res.ok && res.data.exists;
  if (exists) {
    const { unavailable: _unavailable, ...rest } = tab;
    return rest;
  }
  return { ...tab, unavailable: true };
}

/**
 * Rename every extension-less file in the backup folder to `{name}-Corrupted.txt`
 * and remove the corrupt session JSON (UWP `RecoverBackupFilesAsync` rename +
 * fresh start). Idempotent: files that already carry an extension are skipped.
 */
async function renameCorruptedBackups(): Promise<void> {
  const folder = backupFolderPath();
  let entries: string[];
  try {
    entries = await readdir(folder);
  } catch {
    entries = [];
  }
  for (const name of entries) {
    if (name.includes('.')) continue; // skip already-extensioned (e.g. *-Corrupted.txt)
    try {
      await rename(join(folder, name), join(folder, `${name}-Corrupted.txt`));
    } catch {
      // best-effort: a failed rename must not block fresh start
    }
  }
  await rm(sessionFilePath(), { force: true });
  lastSessionJsonStr = null;
}

/**
 * Clear recovered backup files after the user discards recovery (UWP clears the
 * session + backups). Removes the session JSON and every extension-less backup;
 * `{name}-Corrupted.txt` dumps are left for the user to inspect/delete.
 */
export async function clearRecovered(): Promise<Result<void>> {
  try {
    await rm(sessionFilePath(), { force: true });
    const folder = backupFolderPath();
    let entries: string[];
    try {
      entries = await readdir(folder);
    } catch {
      entries = [];
    }
    for (const name of entries) {
      if (name.includes('.')) continue; // preserve *-Corrupted.txt recovery dumps
      await rm(join(folder, name), { force: true });
    }
    lastSessionJsonStr = null;
    return { ok: true, data: undefined };
  } catch (e) {
    return { ok: false, error: errMsg(e) };
  }
}

/**
 * Read a tab's recovered content for the renderer. Prefers Pending (unsaved
 * edits, case (c)) over LastSaved (case (b)); returns null when neither backup
 * exists. Exposed so the loadLast IPC path / renderer can hydrate buffers.
 */
export async function readTabBackup(editorId: string): Promise<TabBackup | null> {
  const folder = backupFolderPath();
  const out: TabBackup = {};
  out.lastSaved = await readBackupFile(join(folder, lastSavedBackupName(editorId)));
  out.pending = await readBackupFile(join(folder, pendingBackupName(editorId)));
  if (out.lastSaved == null && out.pending == null) return null;
  return out;
}

async function readBackupFile(path: string): Promise<string | undefined> {
  try {
    await stat(path);
    return await readFile(path, 'utf8');
  } catch {
    return undefined;
  }
}

/** Test seam: reset the in-memory dirty-check cache (UWP `_lastSessionJsonStr = null`). */
export function __resetSessionDirtyCache(): void {
  lastSessionJsonStr = null;
}

File: src-tauri/src/session.rs
//! session — port of src/main/session.ts (task #3, owner: worker-persist).
//!
//! To preserve: NotepadsSessionData.json v1 + BackupFiles/{id}-LastSaved /
//! -Pending (extensionless); dirty-check vs last JSON (case-insensitive);
//! corrupt → rename *-Corrupted.txt + delete JSON; loadLast re-stats paths →
//! unavailable:true; clearRecovered keeps Corrupted files.

use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;
use std::sync::Mutex;

use serde::{Deserialize, Serialize};
use tauri::Manager;

use crate::contract::{SessionSnapshot, SessionTab, SnapshotResult};
use crate::result::NpResult;

const SESSION_FILE_NAME: &str = "NotepadsSessionData.json";
const BACKUP_FOLDER_NAME: &str = "BackupFiles";

// ---------------------------------------------------------------------------
//  Internal types
// ---------------------------------------------------------------------------

/// Per-tab backup content kept ONLY in main; never serialized into the JSON.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub(crate) struct TabBackupData {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    last_saved: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pending: Option<String>,
}

/// Ordered session JSON shape (version first, matching UWP serialization).
#[derive(Serialize)]
struct SessionFileFormat<'a> {
    version: u32,
    tabs: &'a [SessionTab],
    #[serde(skip_serializing_if = "Option::is_none")]
    active_editor_id: &'a Option<String>,
}

// ---------------------------------------------------------------------------
//  Path helpers
// ---------------------------------------------------------------------------

fn app_data_dir(app: &tauri::AppHandle) -> PathBuf {
    if let Ok(ov) = std::env::var("NOTEPADS_E2E_USERDATA") {
        if !ov.is_empty() {
            return PathBuf::from(ov);
        }
    }
    app.path()
        .app_data_dir()
        .expect("app_data_dir should exist")
}

fn session_file_path(app: &tauri::AppHandle) -> PathBuf {
    app_data_dir(app).join(SESSION_FILE_NAME)
}

fn backup_folder_path(app: &tauri::AppHandle) -> PathBuf {
    app_data_dir(app).join(BACKUP_FOLDER_NAME)
}

fn last_saved_backup_name(editor_id: &str) -> String {
    format!("{editor_id}-LastSaved")
}

fn pending_backup_name(editor_id: &str) -> String {
    format!("{editor_id}-Pending")
}

// ---------------------------------------------------------------------------
//  In-memory dirty-check cache (mirrors UWP _lastSessionJsonStr)
// ---------------------------------------------------------------------------

/// Last serialized session JSON string, used for dirty-check.
static LAST_SESSION_JSON: Mutex<Option<String>> = Mutex::new(None);

fn last_json_str() -> Option<String> {
    LAST_SESSION_JSON.lock().unwrap_or_else(|p| p.into_inner()).clone()
}

fn set_last_json_str(s: Option<String>) {
    *LAST_SESSION_JSON.lock().unwrap_or_else(|p| p.into_inner()) = s;
}

// ---------------------------------------------------------------------------
//  Serialization
// ---------------------------------------------------------------------------

/// Serialize a snapshot to the on-disk JSON string. Only the contract surface
/// is written (version, tabs, activeEditorId). The `_backups` sidecar is
/// excluded — it goes into per-tab backup files.
fn serialize_for_disk(data: &SessionSnapshot) -> String {
    let ordered = SessionFileFormat {
        version: 1,
        tabs: &data.tabs,
        active_editor_id: &data.active_editor_id,
    };
    serde_json::to_string_pretty(&ordered).unwrap_or_default()
}

/// Check if two session JSON strings are equal (case-insensitive, matching
/// UWP `StringComparison.OrdinalIgnoreCase`).
fn session_json_equals(a: &str, b: &str) -> bool {
    a.to_lowercase() == b.to_lowercase()
}

// ---------------------------------------------------------------------------
//  Backup file operations
// ---------------------------------------------------------------------------

/// Write per-tab backup files. LastSaved is written whenever provided.
/// Pending is written only when the tab is dirty and pending text exists.
/// If the tab is clean, any stale pending backup is removed.
fn write_backups(
    folder: &std::path::Path,
    tabs: &[SessionTab],
    backups: &HashMap<String, TabBackupData>,
) -> Result<(), String> {
    fs::create_dir_all(folder).map_err(|e| format!("Failed to create backup dir: {e}"))?;

    // Build a lookup of editorId → isModified
    let modified: HashMap<&str, bool> = tabs.iter().map(|t| (t.editor_id.as_str(), t.is_modified)).collect();

    for (editor_id, backup) in backups {
        if let Some(last_saved) = &backup.last_saved {
            let path = folder.join(last_saved_backup_name(editor_id));
            fs::write(&path, last_saved)
                .map_err(|e| format!("Failed to write backup {path:?}: {e}"))?;
        }

        let pending_path = folder.join(pending_backup_name(editor_id));
        let is_dirty = modified.get(editor_id.as_str()).copied().unwrap_or(false);
        if is_dirty {
            if let Some(pending) = &backup.pending {
                fs::write(&pending_path, pending)
                    .map_err(|e| format!("Failed to write pending backup: {e}"))?;
            } else {
                // Dirty but no pending text → remove stale pending file
                let _ = fs::remove_file(&pending_path);
            }
        } else {
            // Clean tab → remove any stale pending backup
            let _ = fs::remove_file(&pending_path);
        }
    }
    Ok(())
}

/// Delete extension-less backup files whose editorId is no longer in the
/// live session. Files WITH an extension (e.g. *-Corrupted.txt) are skipped.
fn delete_orphaned_backups(folder: &std::path::Path, tabs: &[SessionTab]) {
    let live_names: std::collections::HashSet<String> = tabs
        .iter()
        .flat_map(|t| {
            vec![
                last_saved_backup_name(&t.editor_id),
                pending_backup_name(&t.editor_id),
            ]
        })
        .collect();

    let entries = match fs::read_dir(folder) {
        Ok(iter) => iter,
        Err(_) => return, // folder absent → nothing to prune
    };

    for entry in entries.flatten() {
        let name = entry.file_name();
        let name_str = name.to_string_lossy();
        // Skip files with extension (e.g. *-Corrupted.txt)
        if name_str.contains('.') {
            continue;
        }
        // Skip known-live backups
        if live_names.contains(name_str.as_ref()) {
            continue;
        }
        let _ = fs::remove_file(entry.path());
    }
}

// ---------------------------------------------------------------------------
//  Corruption recovery
// ---------------------------------------------------------------------------

/// Rename every extension-less file in the backup folder to
/// `{name}-Corrupted.txt` and remove the corrupt session JSON.
fn rename_corrupted_backups(app: &tauri::AppHandle) {
    let folder = backup_folder_path(app);
    let entries = match fs::read_dir(&folder) {
        Ok(iter) => iter,
        Err(_) => {
            // No backup folder — just delete the JSON
            let _ = fs::remove_file(session_file_path(app));
            set_last_json_str(None);
            return;
        }
    };

    for entry in entries.flatten() {
        let name = entry.file_name();
        let name_str = name.to_string_lossy();
        // Skip files that already have an extension (e.g. *-Corrupted.txt)
        if name_str.contains('.') {
            continue;
        }
        let corrupted_name = folder.join(format!("{name_str}-Corrupted.txt"));
        let _ = fs::rename(entry.path(), &corrupted_name);
    }

    let _ = fs::remove_file(session_file_path(app));
    set_last_json_str(None);
}

// ---------------------------------------------------------------------------
//  Tauri commands
// ---------------------------------------------------------------------------

/// Persist a session snapshot. Dirty-checked against the last written JSON
/// (case-insensitive). On write: (1) write versioned JSON, (2) write per-tab
/// backups, (3) delete orphaned backups for editors no longer in the session.
///
/// The `backups` parameter is a map from editorId to {lastSaved?, pending?}.
/// Only Pending text is written when the tab is dirty.
#[tauri::command]
pub async fn session_snapshot(
    app: tauri::AppHandle,
    data: SessionSnapshot,
    #[allow(unused_mut)] backups: Option<HashMap<String, TabBackupData>>,
) -> NpResult<SnapshotResult> {
    let session_json_str = serialize_for_disk(&data);

    // Dirty check
    if let Some(ref last) = last_json_str() {
        if session_json_equals(last, &session_json_str) {
            return NpResult::Ok(SnapshotResult { written: false });
        }
    }

    let folder = backup_folder_path(&app);

    // Write session JSON atomically
    let session_path = session_file_path(&app);
    let pid = std::process::id();
    let tmp_path = session_path.with_file_name(format!("{SESSION_FILE_NAME}.{pid}.tmp"));

    if let Some(parent) = tmp_path.parent() {
        if let Err(e) = fs::create_dir_all(parent) {
            return NpResult::Err(format!("Failed to create session dir: {e}"));
        }
    }

    if let Err(e) = fs::write(&tmp_path, &session_json_str) {
        return NpResult::Err(format!("Failed to write session: {e}"));
    }
    if let Err(e) = fs::rename(&tmp_path, &session_path) {
        let _ = fs::remove_file(&tmp_path);
        return NpResult::Err(format!("Failed to persist session: {e}"));
    }

    // Write per-tab backups
    if let Some(ref backups) = backups {
        if let Err(e) = write_backups(&folder, &data.tabs, backups) {
            return NpResult::Err(e);
        }
    }

    // Prune orphaned backups
    delete_orphaned_backups(&folder, &data.tabs);

    set_last_json_str(Some(session_json_str));
    NpResult::Ok(SnapshotResult { written: true })
}

/// Load the last persisted session.
///
/// Returns null when: no session file exists, or the JSON is corrupt (in
/// which case backups are renamed to *-Corrupted.txt and the session file
/// is deleted).
///
/// Each tab's filePath is re-validated via fs::metadata; a missing file
/// sets `unavailable: true` with the path preserved.
#[tauri::command]
pub async fn session_load_last(app: tauri::AppHandle) -> NpResult<Option<SessionSnapshot>> {
    let session_path = session_file_path(&app);

    let raw = match fs::read_to_string(&session_path) {
        Ok(r) => r,
        Err(_) => return NpResult::Ok(None), // No file → fresh start
    };

    // Parse and validate
    let parsed: serde_json::Value = match serde_json::from_str(&raw) {
        Ok(v) => v,
        Err(_) => {
            rename_corrupted_backups(&app);
            return NpResult::Ok(None);
        }
    };

    // Validate version & tabs field
    let version = parsed.get("version").and_then(|v| v.as_u64()).unwrap_or(0);
    let has_tabs = parsed.get("tabs").map(|v| v.is_array()).unwrap_or(false);
    if version != 1 || !has_tabs {
        rename_corrupted_backups(&app);
        return NpResult::Ok(None);
    }

    // Deserialize into SessionSnapshot
    let snapshot: SessionSnapshot = match serde_json::from_value(parsed) {
        Ok(s) => s,
        Err(_) => {
            rename_corrupted_backups(&app);
            return NpResult::Ok(None);
        }
    };

    // Re-validate tab paths
    let mut tabs: Vec<SessionTab> = Vec::new();
    for tab in snapshot.tabs {
        tabs.push(revalidate_tab(tab));
    }

    let result = SessionSnapshot {
        version: 1,
        tabs,
        active_editor_id: snapshot.active_editor_id,
    };

    // Cache the serialized form for dirty-check
    set_last_json_str(Some(serialize_for_disk(&result)));

    NpResult::Ok(Some(result))
}

/// Re-validate a single tab's filePath. Untitled buffers (no path) pass
/// through. Missing files are marked unavailable with path preserved.
fn revalidate_tab(tab: SessionTab) -> SessionTab {
    let Some(ref file_path) = tab.file_path else {
        // Untitled buffer — no path to revalidate
        let mut t = tab;
        t.unavailable = None;
        return t;
    };

    let exists = std::path::Path::new(file_path).exists();

    if exists {
        let mut t = tab;
        t.unavailable = None;
        t
    } else {
        SessionTab {
            unavailable: Some(true),
            ..tab
        }
    }
}

/// Clear recovered backup files. Removes the session JSON and every
/// extension-less backup. `{name}-Corrupted.txt` dumps are preserved
/// for user inspection.
#[tauri::command]
pub async fn session_clear_recovered(app: tauri::AppHandle) -> NpResult<()> {
    // Remove session JSON
    let _ = fs::remove_file(session_file_path(&app));

    let folder = backup_folder_path(&app);
    let entries = match fs::read_dir(&folder) {
        Ok(iter) => iter,
        Err(_) => {
            set_last_json_str(None);
            return NpResult::Ok(());
        }
    };

    for entry in entries.flatten() {
        let name = entry.file_name();
        let name_str = name.to_string_lossy();
        // Preserve files with extension (Corrupted recovery dumps)
        if name_str.contains('.') {
            continue;
        }
        let _ = fs::remove_file(entry.path());
    }

    set_last_json_str(None);
    NpResult::Ok(())
}

/// Test seam: reset the in-memory dirty-check cache.
#[cfg(test)]
pub fn __reset_session_dirty_cache() {
    set_last_json_str(None);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_serialize_for_disk_excludes_backups() {
        let snapshot = SessionSnapshot {
            version: 1,
            tabs: vec![SessionTab {
                editor_id: "ed-1".into(),
                file_path: Some("C:/notes.txt".into()),
                encoding_id: "UTF-8".into(),
                eol_id: crate::contract::EolId::Crlf,
                is_modified: false,
                selection_start: 0.0,
                selection_end: 0.0,
                scroll_top: 0.0,
                view_mode: crate::contract::ViewMode { preview: false, diff: false },
                unavailable: None,
            }],
            active_editor_id: Some("ed-1".into()),
        };

        let json = serialize_for_disk(&snapshot);
        let parsed: serde_json::Value = serde_json::from_str(&json).unwrap();

        assert_eq!(parsed["version"], serde_json::json!(1));
        assert!(parsed["tabs"].is_array());
        // Backups must NOT be present in the persisted JSON
        assert!(parsed.get("backups").is_none());
        assert!(parsed.get("_backups").is_none());
    }

    #[test]
    fn test_session_json_equals_case_insensitive() {
        let a = r#"{"version":1,"tabs":[],"activeEditorId":null}"#;
        let b = r#"{"VERSION":1,"TABS":[],"ACTIVEEDITORID":null}"#;
        assert!(session_json_equals(a, b));
    }

    #[test]
    fn test_revalidate_tab_untitled() {
        let tab = SessionTab {
            editor_id: "ed-1".into(),
            file_path: None,
            encoding_id: "UTF-8".into(),
            eol_id: crate::contract::EolId::Crlf,
            is_modified: false,
            selection_start: 0.0,
            selection_end: 0.0,
            scroll_top: 0.0,
            view_mode: crate::contract::ViewMode { preview: false, diff: false },
            unavailable: None,
        };
        let result = revalidate_tab(tab);
        assert_eq!(result.unavailable, None);
        assert_eq!(result.file_path, None);
    }

    #[test]
    fn test_revalidate_tab_missing_file() {
        let tab = SessionTab {
            editor_id: "ed-1".into(),
            file_path: Some("Z:/nonexistent/file.txt".into()),
            encoding_id: "UTF-8".into(),
            eol_id: crate::contract::EolId::Crlf,
            is_modified: false,
            selection_start: 0.0,
            selection_end: 0.0,
            scroll_top: 0.0,
            view_mode: crate::contract::ViewMode { preview: false, diff: false },
            unavailable: None,
        };
        let result = revalidate_tab(tab);
        assert_eq!(result.unavailable, Some(true));
        assert_eq!(result.file_path, Some("Z:/nonexistent/file.txt".into()));
    }

    #[test]
    fn test_last_saved_backup_name() {
        assert_eq!(last_saved_backup_name("editor-123"), "editor-123-LastSaved");
    }

    #[test]
    fn test_pending_backup_name() {
        assert_eq!(pending_backup_name("editor-456"), "editor-456-Pending");
    }
}
